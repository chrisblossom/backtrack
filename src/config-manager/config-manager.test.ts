import path from 'path';

function configManager<T>(args: { config: T; namespace: string }): T {
	const Backtrack = require('../initialize/initialize').Initialize;

	const backtrack = new Backtrack();

	return backtrack.configManager(args);
}

describe('configManager', () => {
	const cwd = process.cwd();

	beforeEach(() => {
		jest.mock('../utils/log', () => ({
			warn: jest.fn(),
			info: jest.fn(),
			error: jest.fn(),
			success: jest.fn(),
		}));
	});

	afterEach(() => {
		process.chdir(cwd);
	});

	test('throws when backtrack config is not found', () => {
		const dir = path.resolve(
			__dirname,
			'__sandbox__/backtrack-config-missing/',
		);
		process.chdir(dir);
		jest.doMock('parent-module', () => () => path.join(dir, 'testing.js'));

		const config = {
			extends: ['airbnb-base'],
		};

		let error;
		try {
			configManager({
				config,
				namespace: 'backtrack-config-missing',
			});
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchInlineSnapshot(
				`[Error: backtrack config not found]`,
			);
		}
	});

	test('returns original config when no custom config found for current config', () => {
		const dir = path.resolve(__dirname, '__sandbox__/no-custom-config-2/');
		process.chdir(dir);
		jest.doMock('parent-module', () => () => path.join(dir, 'testing.js'));

		const config = {
			extends: ['airbnb-base'],
		};

		const result = configManager({
			config,
			namespace: 'no_custom_config_2',
		});

		expect(result === config).toEqual(true);
	});

	test('merges config by default', () => {
		const dir = path.resolve(__dirname, '__sandbox__/custom-config-1/');
		process.chdir(dir);
		jest.doMock('parent-module', () => () => path.join(dir, 'testing.js'));

		const config = {
			extends: ['airbnb-base'],
		};

		const result = configManager({
			config,
			namespace: 'custom_config_1',
		});

		expect(result).toEqual({
			extends: ['airbnb-base'],
			rules: { semi: 'always' },
		});
	});

	test('handles nested config file', () => {
		const dir = path.resolve(__dirname, '__sandbox__/custom-config-1/');
		process.chdir(dir);
		jest.doMock(
			'parent-module',
			() => () => path.join(dir, 'nested/testing.js'),
		);

		const config = {
			extends: ['airbnb-base'],
		};

		const result = configManager({
			config,
			namespace: 'custom_config_1',
		});

		expect(result).toEqual({
			extends: ['airbnb-base'],
			rules: { semi: 'always' },
		});
	});
});
