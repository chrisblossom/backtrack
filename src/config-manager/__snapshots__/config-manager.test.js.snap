// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configManager throws when backtrack config is not found 1`] = `
Object {
  "exitCode": 1,
  "message": "backtrack config not found",
}
`;
